scripts: west: runners: nrfjprog: add basic support for nRF54H series

gmarull opened the pull request:
Note: I see UICR merged hex is still not generated, however, I've tested with a hex from -next and it seems to digest it well. I wonder if there's a real need for a merged hex, we could likely invoke nrfjprog twice (one for UICR, other for regular zephyr.hex).

anangl committed:
drivers: serial: nrfx: Allow new UARTE instances to be used

Extend Kconfig definitions and nrfx_config translations so that UARTE
instances that are available in nRF54H20 can be used.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: Add initial support for nRF54H20 Eng0 SoC

Add definition of the nRF54H20 SoC revision Eng0 with its Application,
Radio, and Peripheral Processor (PPR) cores and basic peripherals:
GRTC, GPIOs, and UARTs.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
soc: nordic: Add initial support for nRF54H20 Eng0

Add nrfx and Kconfig related infrastructure plus SoC initialization
code to allow building for nRF54H20 targets (Application and Radio
cores).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: arm: Add initial support for nRF54H20 PDK

Add targets that allows building for the Application and Radio cores
in the nRF54H20 SoC on the nRF54H20 PDK board.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull committed:
dt-bindings: misc: add nordic,nrf-ficr helper definitions

Add definitions to access some FICR fields, for now HSFLL trim values.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: nrf54h20pdk_nrf54h20: enable nrfjprog runner

So that we can flash the board.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: west: runners: nrfjprog: add option to not erase

One some new SoCs, the erase option is not required, so introduce a new
option that doesn't set any erase mode.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: west: runners: nrfjprog: add erasepage op

Add a new operation that allows erasing pages.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: west: runners: nrfjprog: add basic support for nRF54H series

Add basic support to flash application and/or radio core for nRF54H
series. Note that features like merged hexes present in nRF53 series is
not supported. Flashing other cores like PPR/VPR is not supported yet.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented:
cc: @57300

57300 commented on scripts/west_commands/runners/nrf_common.py:
```diff
@@ -260,15 +273,35 @@ def program_hex(self):
+        elif self.build_conf.getboolean('CONFIG_SOC_NRF54H20_ENG0_CPURAD'):
+            core = 'NRFDL_DEVICE_CORE_NETWORK'
+        else:
+            raise NotImplementedError("Unsupported core")
```
I don't know how nrfjprog works, but in theory, choosing between app and radio shouldn't matter much when flashing, because neither domain has write access to MRAM by default. I think this is mediated by secure domain in either case.

I wonder if it works to flash everything via `CP_APPLICATION`. That would simplify supporting PPR.

gmarull replied:
This is the domain to core table we have in -next as of today:

```
_DOM2COP = {
    "application": "CP_APPLICATION",
    "radiocore": "CP_NETWORK",
    "ppr": "CP_APPLICATION",
    "secure": "CP_SECURE",
    "sysctrl": "CP_APPLICATION",
}
```

So as you can see, ppr for example uses CP_APPLICATION. I'm not sure if core can be obviated or not, probably a question for nrfjprog authors?

57300 replied:
Based on my own testing, omitting `--coprocessor` is fine in common cases. I suggest doing that, and removing the `NotImplementedError`. For special cases, we can add this out of tree:
```cmake
board_runner_args(nrfjprog "--tool-opt=--coprocessor (...)")
```

gmarull replied:
thanks for the investigation, it looks like nrf54h20 flashing gets easier from the initial "specs". I've simplified the patch, and checked with the latest changes, cpuapp boots fine and UICR is erased/flashed.

57300 commented on scripts/west_commands/runners/nrf_common.py:
```diff
@@ -260,15 +273,35 @@ def program_hex(self):
+        else:
+            raise NotImplementedError("Unsupported core")
+
+        if self.hex_has_uicr_content() and not self.erase_uicrs:
+            self.exec_op('erasepage', defer=True, core=core,
+                         page=UICR_RANGES[self.family][core][0])
```
Kind of an edge case, but if app's hex somehow contains radio's UICR, then only app's UICR will be erased.

gmarull replied:
I'd say it's a case we should avoid? To generate the radiocore UICR we'd need a cpurad build as well, so we'll have the cpurad image as well, right?

gmarull replied:
Btw, I've done it this way in case you just rebuild cpuapp, but radiocore stays the same (and there's no resource overlap in the changes...). I added `--erase-uicrs` CLI option to force the erase of all UICRs

57300 replied:
Sure, it's just something we may have to be conscious of in the context of sysbuild and especially in NCS.

57300 reviewed (commented):
> Note: I see UICR merged hex is still not generated

@anangl hasn't added the code yet, but it should generate a `uicr_merged.hex` and update `runners.yaml` accordingly.

I'm also not sure if it's necessary, but it seems more robust than some other alternatives, like caching the path to `uicr.hex` in CMake, or hardcoding `zephyr_hex.parent / "uicr.hex"` in the runner.

57300 committed:
modules: hal_nordic: Integrate nrf-regtool

nrf-regtool is a Python utility from Nordic Semiconductor, which is used
for generating binary files with register values for given peripherals.
It sources the descriptions of peripheral registers from CMSIS-SVD files
(typically ones bundled with nRF MDK).

For some peripherals, such as UICR, nrf-regtool supports parsing values
from devicetree as well, based on the bindings already found in Zephyr.

Currently, this tool is not submitted as a script to Zephyr, but it can
be installed from PyPI.

Having nrf-regtool installed is recommended when working with nRF54H20.
Booting the Application or Radiocore CPU requires flashing not only its
firmware, but also its respective UICR instance. On this SoC, the UICR
is used to assign ownership of global hardware resources, including
memory and peripherals, to individual cores. The Zephyr build system can
call nrf-regtool to generate the UICR based on devicetree, to reflect
the boot-time hardware configuration required for a given application.
The generated `uicr.hex` is then merged with `zephyr.hex`, so that they
can be flashed together using west.

The build system integration takes the form of a CMake package, which
includes a version check and reusable components; over time, some of
these components can be reused by sysbuild. This package is located in
the `hal_nordic` module, because it depends on the `SOC_SVD_FILE` CMake
variable, which is defined there as well.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
modules: hal_nordic: Set SOC_SVD_FILE

This CMake variable contains a path to a CMSIS-SVD file from nRF MDK,
which describes the selected Nordic SoC.

Initially, this will be used as an argument to nrf-regtool.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

gmarull commented on scripts/west_commands/runners/nrf_common.py:
```diff
@@ -266,11 +277,25 @@ def program_hex(self):
+            for start_addr, _ in UICR_RANGES['NRF54H_FAMILY']:
+                self.exec_op('erasepage', defer=True, page=start_addr)
+        elif uicr_addrs:
+            for uicr_addr in uicr_addrs:
+                self.exec_op('erasepage', defer=True, page=uicr_addr)
```
@57300 this will erase all UICRs in the hex file (we could now have app+app_uicr+rad_uicr)

57300 commented on scripts/west_commands/runners/nrf_common.py:
```diff
@@ -213,7 +224,7 @@ def flush(self, force=False):
                 # verify failed du to the UICR not been erased before, so giving
                 # a warning here will hopefully enhance UX.
-                if self.hex_has_uicr_content():
+                if self.hex_get_uicrs():
```
How about adding this for readability:
```py
def hex_has_uicr_content(self):
    return bool(self.hex_get_uicrs())
```

57300 commented on scripts/west_commands/runners/nrf_common.py:
```diff
@@ -266,11 +277,25 @@ def program_hex(self):
+    def program_hex_nrf54h(self):
+        uicr_addrs = self.hex_get_uicrs()
+
+        if self.erase_all_uicrs:
+            for start_addr, _ in UICR_RANGES['NRF54H_FAMILY']:
+                self.exec_op('erasepage', defer=True, page=start_addr)
+        elif uicr_addrs:
+            for uicr_addr in uicr_addrs:
+                self.exec_op('erasepage', defer=True, page=uicr_addr)
```
I would prefer a single loop:
```py
if self.erase_all_uicrs:
    uicr_addrs = UICR_RANGES['NRF54H_FAMILY']
else:
    uicr_addrs = self.hex_get_uicrs()

for start_addr, _ in uicr_addrs:
    self.exec_op('erasepage', defer=True, page=start_addr)
```

This would require updating `hex_get_uicrs()` with e.g.:
```py
uicrs = []
for region in UICR_RANGES[self.family]:
    if self.hex_refers_region(*region):
        uicrs.append(region)
```

gmarull replied:
done

57300 reviewed (approved):
Works for me. Just some minor suggestions.

anangl committed:
tests: arm_irq_vector_table: Add special handling for nRF54H20

Different set of IRQ lines need to be used for this SoC and the CLOCK
IRQ is not to be installed in the vector table.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
I tried to flash hello_world built for `nrf54h20pdk_nrf54h20_cpurad` and got the following error:
```
Address 0x0FFFA000 does not map to a known memory.
ERROR: Argument provided has a wrong value. Memory access not aligned, value
ERROR: outside of range or similar.
NOTE: For additional output, try running again with logging enabled (--log).
NOTE: Any generated log error messages will be displayed.
FATAL ERROR: command exited with status 54: nrfjprog --erasepage 0x0fffa000 -f NRF54H --snr 1050886495
```

gmarull commented:
> I tried to flash hello_world built for `nrf54h20pdk_nrf54h20_cpurad` and got the following error:
> 
> ```
> Address 0x0FFFA000 does not map to a known memory.
> ERROR: Argument provided has a wrong value. Memory access not aligned, value
> ERROR: outside of range or similar.
> NOTE: For additional output, try running again with logging enabled (--log).
> NOTE: Any generated log error messages will be displayed.
> FATAL ERROR: command exited with status 54: nrfjprog --erasepage 0x0fffa000 -f NRF54H --snr 1050886495
> ```

can you try passing `--tool-opt="--coprocessor CP_NETWORK"`?

anangl commented:
> can you try passing `--tool-opt="--coprocessor CP_NETWORK"`?

Yes. with this added, it worked fine.

gmarull commented:
> > can you try passing `--tool-opt="--coprocessor CP_NETWORK"`?
> 
> Yes. with this added, it worked fine.

@57300 so it looks like `--coprocessor` cannot be entirely obviated... I think we should just play safe and add it when known.

57300 commented:
> > > can you try passing `--tool-opt="--coprocessor CP_NETWORK"`?
> > 
> > 
> > Yes. with this added, it worked fine.
> 
> @57300 so it looks like `--coprocessor` cannot be entirely obviated... I think we should just play safe and add it when known.

Indeed. I didn't think to test `--erasepage` earlier, and that's the command causing the problem, in that it requires a specific coprocessor for some reason.

anangl committed:
drivers: serial: Kconfig.nrfx: Filter out options unsupported on nRF54H20

On nRF54H20, only the new shim can be used and the enhanced poll out
cannot be enabled since there is no DPPI support for this SoC yet.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: timer: nrf_grtc_timer: Add dependency on nRF clock control

... in the related parts, so that the driver can be used on nRF54H20
where the clock control is not present yet.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: lib: cpp: Exclude cpp98 test on nRF54H20

Similarly to nRF54L15, the MDK files in HAL for this SoC are not
compatible with C++98, so currently this test cannot be performed.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull commented:
@anangl please try now

anangl reviewed (approved):
Works fine now.

gmarull commented:
integrated
